### PR TITLE
Support Airplay 1 and 2 (with a single Airplay 2)

### PR DESCRIPTION
**Pull request by linknum23**
This is a continuation of the work started in #378 
- [x]  Add shairport-sync binary built for aiplay2
- [x]  Add nqptp dependency (needs to be installed from source for now)
- [x]  Test single airplay 2 stream with additional airplay 1 streams
- [x]  Add support for single airplay 2 stream, with the second stream either being rejected or converted to airplay 1 (depends on above)
- [x]  Add metadata using MPRIS support
- [x] Resolve MPRIS shutdown delay

### CODE REVIEW

**Comment by linknum23** [comment]
We are still finalizing the integration with the metadata using the MPRIS interface. The MPRIS interface is not shutting down well yet.

File: streams/mprischildproc.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    if mpris:
+      try:
+        print('disconnecting from MPRIS proxy', flush=True)
+        disconnect_proxy(mpris)
+      except Exception as e:
+        print(e, flush=True)
+
+
+if len(sys.argv) != 3:
+  print("Usage: mprisdaughterproc.py <service_suffix> <metadata_path>")
```
Comment (linknum23) [inline streams/mprischildproc.py:129]: I know this interface is fairly simple, but does it make sense to use argparse here?

File: streams/mprischildproc.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            object_path = "/org/mpris/MediaPlayer2",
+            interface_name = "org.mpris.MediaPlayer2.Player"
+          )
+        except Exception as e:
+          metadata['connected'] = False
+          print(f"failed to connect mpris {e}", flush=True)
+          if not self.ok:
+            break
+
+      #print(f"getting mrpis metadata from {self.service_suffix}")
```
Comment (linknum23) [inline streams/mprischildproc.py:67]: Do you want to leave in a debug mode accessible by a flag?

File: streams/mprischildproc.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+
```
Comment (linknum23) [inline streams/mprischildproc.py:1]: Suggestion for filename mpris_metadata_reader

Comment (linknum23) [reply streams/mprischildproc.py:1]: Running this with pylint: 
```bash
source venv/bin/activate
pylint streams/mprischildproc.py
```
```
streams/mprischildproc.py:106:0: C0301: Line too long (101/100) (line-too-long)
streams/mprischildproc.py:19:0: C0115: Missing class docstring (missing-class-docstring)
streams/mprischildproc.py:19:0: C0103: Class name "mpris_child_proc" doesn't conform to PascalCase naming style (invalid-name)
streams/mprischildproc.py:38:4: C0103: Attribute name "ok" doesn't conform to snake_case naming style (invalid-name)
streams/mprischildproc.py:20:21: W0621: Redefining name 'service_suffix' from outer scope (line 132) (redefined-outer-name)
streams/mprischildproc.py:20:37: W0621: Redefining name 'metadata_path' from outer scope (line 133) (redefined-outer-name)
streams/mprischildproc.py:48:2: R0912: Too many branches (22/12) (too-many-branches)
streams/mprischildproc.py:48:2: R0915: Too many statements (58/50) (too-many-statements)
streams/mprischildproc.py:132:0: C0103: Constant name "service_suffix" doesn't conform to UPPER_CASE naming style (invalid-name)
```

File: scripts/configure.py [hunk 4]
```diff
@@ -122,8 +122,21 @@
...
   'airplay' : {
-    'apt' : [ 'shairport-sync' ],
-    'copy' : [{'from': 'bin/ARCH/shairport-sync-metadata-reader', 'to': 'streams/shairport-sync-metadata-reader'}],
+    'apt' : [ 'shairport-sync'],
+    'copy' : [{'from': 'bin/ARCH/shairport-sync-ap2', 'to': 'streams/shairport-sync-ap2'},
+              {'from': 'bin/ARCH/shairport-sync', 'to': 'streams/shairport-sync'}],
+    'script': [
+        'if which nqptp  > /dev/null; then exit 0; fi',
+        'cd ~',
+        'git clone https://github.com/mikebrady/nqptp.git',
```
Comment (linknum23) [inline scripts/configure.py:131]: Suggestion: use a temporary directory

File: bin/generate_bins.bash [hunk 5]
```diff
@@ -4,6 +4,28 @@ set -e # stop on error
...
 
+
+helptext="Usage: generate_bins.bash [OPTION]...
+Build binaries that are packaged with AmpliPi for several different architectures.
+This builds binaries on a Pi for now, TODO: cross compile everything!
+
+  --only-local: Only build local binaries (the pi takes awhile)
+  -h, --help: Print this help text.
+"
+
```
[the remark this answers is not in the document]
Comment (linknum23) [reply bin/generate_bins.bash:15]: Not needed for this PR!

File: amplipi/streams.py [hunk 6]
```diff
@@ -208,9 +216,22 @@ def connect(self, src):
     """ Connect an AirPlay device to a given audio source
     This creates an AirPlay streaming option based on the configuration
     """
+
+    # if stream is airplay2 check for other airplay2s and error if found
+    # pgrep has it's own process that will include the process name so we sub 1 from the results
+    if self.ap2:
+      if len(os.popen("pgrep -f shairport-sync-ap2").read().strip().splitlines())-1 > 0:
+        self.ap2_exists = True
+        self._connect(src)
```
Comment (linknum23) [inline amplipi/streams.py:225]:
```suggestion
        print('Another Airplay 2 stream is already in use, unable to start, mocking connection'.')
        self._connect(src)
```

File: amplipi/streams.py [hunk 7]
```diff
@@ -234,54 +255,115 @@ def connect(self, src):
...
+          img_loc = f'generated/{self.src}/{img_name}'
+          source.img_url = img_loc
+        else:
+          source.track = "No metadata available"
+
+
+    except Exception as e:
+      print(f"error in airplay: {e}")
+
+    print(f'airplay {self.name} returns {source.__dict__}')
```
Comment (linknum23) [inline amplipi/streams.py:347]: This is debugging info, should we add a debug flag to the info method?

**Review by linknum23 — COMMENTED** [review]
Initial comments. Some changes needed.

File: amplipi/streams.py [hunk 8]
```diff
@@ -234,54 +255,115 @@ def connect(self, src):
...
-    except Exception as exc:
-      print(f'Failed to get currentSong - it may not exist: {exc}')
+      md = self.mpris.metadata()
+
+      if self.mpris.is_playing():
+        source.state = 'playing'
+      else:
+        # if we've been paused for a while, say we're stopped since
+        # shairport-sync doesn't really differentiate between paused and stopped
+        if source.state == 'paused' and time.time() - md.state_changed_time < self.STATE_TIMEOUT:
```
Comment (linknum23) [inline amplipi/streams.py:325]: This logic doesn't appear to actually work

File: amplipi/streams.py [hunk 9]
```diff
@@ -219,12 +240,12 @@ def connect(self, src):
...
         'resync_threshold': 0, # a synchronisation error greater than this will cause resynchronisation; 0 disables it
         'log_verbosity': 0, # "0" means no debug verbosity, "3" is most verbose.
+        'mpris_service_bus': 'Session',
       },
       'metadata':{
         'enabled': 'yes',
         'include_cover_art': 'yes',
-        'pipe_name': f'{utils.get_folder("config")}/srcs/{src}/shairport-sync-metadata',
-        'pipe_timeout': 5000,
+        'cover_art_cache_directory': f'{utils.get_folder("web")}/generated/{src}',
```
Comment (linknum23) [inline amplipi/streams.py:248]: Should this directory match `web_dir`?

File: amplipi/streams.py [hunk 10]
```diff
@@ -234,54 +255,115 @@ def connect(self, src):
...
+          source.state = 'stopped'
+
+      if source.state != 'stopped':
+        source.artist = md.artist
+        source.track = md.title
+        source.album = md.album
+        source.supported_cmds=list(self.supported_cmds)
+
+        if md.title != '':
+          img_name = os.listdir(f'{utils.get_folder("web")}/generated/{self.src}')[0]
```
Comment (linknum23) [inline amplipi/streams.py:337]: This throws an exception if the firectory is empty which seems like a pretty normal case

File: amplipi/streams.py (hunk 9 above)
Comment (linknum23) [reply amplipi/streams.py:248]: this path is used in a bunch of places, make it class var?

File: README.md [hunk 11]
```diff
@@ -22,7 +22,7 @@ Most of these digital streaming services are provided by other open-source proje
 |Streaming Service|Supported|Multiple?|Metadata|Provided By|Notes|
 |--|--:|--:|--:|--|--|
 |Pandora|Yes|Yes|Yes|[Pianobar](https://github.com/PromyLOPh/pianobar)||
-|Airplay|Yes|Yes|Yes|[Shairport-sync](https://github.com/mikebrady/shairport-sync)|Metadata only available from iTunes|
+|Airplay|Yes|Yes|Yes|[Shairport-sync](https://github.com/mikebrady/shairport-sync)|Airplay 2 for [one stream only](https://github.com/micro-nova/AmpliPi/issues/436).|
 |Spotify|Yes|Yes|Yes|[Librespot](https://github.com/librespot-org/librespot)|Requires Spotify Premium, one account per stream. See [disclaimer](https://github.com/librespot-org/librespot#disclaimer)|
```
[the remark this answers is not in the document]
Comment (klay2000) [reply README.md:26]: we can, however librespot is the underlying library that both spotifyd and volibrespot use

File: amplipi/ctrl.py [hunk 12]
```diff
@@ -301,18 +301,21 @@ def reinit(self, settings: models.AppSettings = models.AppSettings(), change_not
     self._update_groups()
 
   def __del__(self):
+    print('controller shutting down', flush=True)
```
[the remark this answers is not in the document]
Comment (klay2000) [reply amplipi/ctrl.py:304]: whoops missed the controller ones

File: streams/mprischildproc.py (hunk 1 above)
Comment (klay2000) [reply streams/mprischildproc.py:129]: I think that might be overkill, honestly I'm not sure the print on incorrect args was even necessary given nobody should ever see that if it's only run by Amplipi

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (linknum23) [reply README.md:26]: oh yeah, nevermind

File: amplipi/streams.py (hunk 8 above)
Comment (linknum23) [reply amplipi/streams.py:325]: When playing spotify over airplay the stream is actually capable of restarting playback from a paused state with the play and next commands..

File: streams/mprischildproc.py (hunk 1 above)
Comment (linknum23) [reply streams/mprischildproc.py:129]: LoL it is a bit until you have to debug it. I'd leave the arg printout at least for the next person.

File: amplipi/streams.py (hunk 8 above)
Comment (klay2000) [reply amplipi/streams.py]: try testing this again, I think I fixed it

File: scripts/configure.py (hunk 4 above)
Comment (klay2000) [reply scripts/configure.py:131]: take a look at what I changed it to, not sure if what I did works or now

File: amplipi/streams.py (hunk 9 above)
Comment (linknum23) [reply amplipi/streams.py]: ??

File: amplipi/streams.py [hunk 13]
```diff
@@ -234,54 +261,117 @@ def connect(self, src):
...
+        source.track = md.title
+        source.album = md.album
+        source.supported_cmds=list(self.supported_cmds)
+
+        if md.title != '':
+          #if there is a title, attempt to get coverart
+          try:
+            img_name = os.listdir(self._coverart_dir)[0]
+            img_loc = f'generated/{self.src}/{img_name}'
+            source.img_url = img_loc
```
Comment (linknum23) [inline amplipi/streams.py:347]: This won't throw an exception when cover art is missing
```suggestion
            images = os.listdir(self._coverart_dir)
            if len(images) > 0:
              source.img_url = img_loc = f'generated/{self.src}/{images[0]}'
```

**Review by linknum23 — COMMENTED** [review]
Cover art is not working on amplipi13, which has been updated to the tip of airplay1and2

**Comment by linknum23** [comment]
Hmm still broken, cover art is only showing the stock airplay image:

Error from the logs: 
```error in airplay: [Errno 2] No such file or directory: '/home/pi/amplipi-dev/web/generated/3'```

**Comment by linknum23** [comment]
Here is a relevant log of the startup:
```
Feb 01 11:16:30 amplipi13 authbind[846]: Spotify2 connected to 1
Feb 01 11:16:30 amplipi13 authbind[846]: rm: cannot remove '/home/pi/amplipi-dev/config/srcs/None/*': No such file or directory
Feb 01 11:16:30 amplipi13 authbind[846]: disconnecting airplaypi!
Feb 01 11:16:30 amplipi13 authbind[846]: Error removing airplay config files: Command 'rm -r /home/pi/amplipi-dev/config/srcs/None/*' returned non-zero exit status 1.
Feb 01 11:16:30 amplipi13 authbind[846]: airplaypi disconnected
Feb 01 11:16:30 amplipi13 authbind[846]: rm: cannot remove '/home/pi/amplipi-dev/config/srcs/None/*': No such file or directory
Feb 01 11:16:30 amplipi13 authbind[846]: disconnecting airplaypi!
Feb 01 11:16:30 amplipi13 authbind[846]: Error removing airplay config files: Command 'rm -r /home/pi/amplipi-dev/config/srcs/None/*' returned non-zero exit status 1.
Feb 01 11:16:30 amplipi13 authbind[846]: airplaypi disconnected
Feb 01 11:16:30 amplipi13 authbind[846]: Loading config from "./config.toml"
Feb 01 11:16:30 amplipi13 authbind[846]: No username specified. Checking username_cmd
Feb 01 11:16:30 amplipi13 authbind[846]: No username_cmd specified
Feb 01 11:16:30 amplipi13 authbind[846]: No password specified. Checking password_cmd
Feb 01 11:16:30 amplipi13 authbind[846]: No password_cmd specified
Feb 01 11:16:30 amplipi13 authbind[846]: No proxy specified
Feb 01 11:16:30 amplipi13 authbind[846]: Using software volume controller.
Feb 01 11:16:30 amplipi13 authbind[846]: no usable credentials found, enabling discovery
Feb 01 11:16:30 amplipi13 authbind[846]: shairport_args: ['/home/pi/amplipi-dev/streams/shairport-sync', '-c', '/home/pi/amplipi-dev/config/srcs/2/shairport.conf']
Feb 01 11:16:30 amplipi13 authbind[846]: airplaypi connected to 2
Feb 01 11:16:30 amplipi13 authbind[846]: rm: cannot remove '/home/pi/amplipi-dev/config/srcs/None/*': No such file or directory
Feb 01 11:16:30 amplipi13 authbind[846]: disconnecting airplaypi2!
Feb 01 11:16:30 amplipi13 authbind[846]: Error removing airplay config files: Command 'rm -r /home/pi/amplipi-dev/config/srcs/None/*' returned non-zero exit status 1.
Feb 01 11:16:30 amplipi13 authbind[846]: airplaypi2 disconnected
Feb 01 11:16:30 amplipi13 authbind[846]: rm: cannot remove '/home/pi/amplipi-dev/config/srcs/None/*': No such file or directory
Feb 01 11:16:30 amplipi13 authbind[846]: disconnecting airplaypi2!
Feb 01 11:16:30 amplipi13 authbind[846]: Error removing airplay config files: Command 'rm -r /home/pi/amplipi-dev/config/srcs/None/*' returned non-zero exit status 1.
Feb 01 11:16:30 amplipi13 authbind[846]: airplaypi2 disconnected
Feb 01 11:16:31 amplipi13 authbind[846]: shairport_args: ['/home/pi/amplipi-dev/streams/shairport-sync-ap2', '-c', '/home/pi/amplipi-dev/config/srcs/3/shairport.conf']
Feb 01 11:16:31 amplipi13 authbind[846]: airplaypi2 connected to 3
```

File: scripts/configure.py [hunk 14]
```diff
@@ -122,8 +122,21 @@
...
   },
   'airplay' : {
-    'apt' : [ 'shairport-sync' ],
-    'copy' : [{'from': 'bin/ARCH/shairport-sync-metadata-reader', 'to': 'streams/shairport-sync-metadata-reader'}],
+    'apt' : [ 'shairport-sync'],
+    'copy' : [{'from': 'bin/ARCH/shairport-sync-ap2', 'to': 'streams/shairport-sync-ap2'},
+              {'from': 'bin/ARCH/shairport-sync', 'to': 'streams/shairport-sync'}],
+    'script': [
+        'if which nqptp  > /dev/null; then exit 0; fi',
+        'cd /tmp',
```
Comment (linknum23) [inline scripts/configure.py:130]:
```suggestion
        'pushd $(mktemp --directory)',
```

File: scripts/configure.py [hunk 15]
```diff
@@ -122,8 +122,21 @@
...
+        'git clone https://github.com/mikebrady/nqptp.git',
+        'pushd nqptp',
+        'autoreconf -fi',
+        './configure --with-systemd-startup',
+        'make',
+        'sudo make install',
+        'sudo systemctl enable nqptp && sudo systemctl restart nqptp',
+        'popd',
```
Comment (linknum23) [inline scripts/configure.py:138]:
```suggestion
        'popd',
        'popd',
```

Comment (linknum23) [reply scripts/configure.py:138]: Not really needed but this makes everything consistent

File: scripts/configure.py [hunk 16]
```diff
@@ -122,8 +122,22 @@
...
   },
   'airplay' : {
-    'apt' : [ 'shairport-sync' ],
-    'copy' : [{'from': 'bin/ARCH/shairport-sync-metadata-reader', 'to': 'streams/shairport-sync-metadata-reader'}],
+    'apt' : [ 'shairport-sync'],
+    'copy' : [{'from': 'bin/ARCH/shairport-sync-ap2', 'to': 'streams/shairport-sync-ap2'},
+              {'from': 'bin/ARCH/shairport-sync', 'to': 'streams/shairport-sync'}],
+    'script': [
+        'if which nqptp  > /dev/null; then exit 0; fi',
+        'pushd /tmp',
```
Comment (linknum23) [inline scripts/configure.py:130]:
```suggestion
        'pushd $(mktemp --directory)',
```

**Review by linknum23 — COMMENTED** [review]
:+1:

**Pull request merged**